exec/cnex/number.c

```diff
@@ -17,49 +17,125 @@
 char *number_to_string(Number x)
 {
     static char buf[50];
-    char val[50];
-    bid128_to_string(val, x);
-    sprintf(buf, "%g", atof(val));
+    number_toString(x, buf, sizeof(buf));
     return buf;
 }
 
-/* ToDo: Implement proper number to string formatter. */
 void number_toString(Number x, char *buf, size_t len)
 {
-    //const int PRECISION = 34; /* ToDo: Finish implementation */
-    int iLeadingZeros = 0;
+    memset(buf, 0x00, len);
+    const int PRECISION = 34;
 
     assert(len != 0);
 
-    char val[50];
-    char lead[50];
+    char val[50] = { 0 };
     bid128_to_string(val, x);
 
-    char *s = &val[0];
-
-    /* Skip over the leading +, if there is one. */
-    while(*s != '\0' && (*s == '+')) s++;
+    char *v, *s = v = &val[1];
 
-    char *v = s;
-
-    while (*s != '\0' && (*s != 'E')) s++;
+    // Calculate the length of the string, locating the exponent value at the same time.
+    int slen = 0;
+    while (*s != '\0' && (*s != 'E')) {
+        s++;
+        slen++;
+    }
 
+    // Store the position of the exponent marker.
     char *E = s;
 
-    if (*s == 'E') {
-        /* Deal with the exponent. */
-        int i;
-        iLeadingZeros = abs(atoi(++s));
-        for (i = 0; i < iLeadingZeros; i++) {
-            lead[i] = '0';
-        }
-        lead[i++] = '.';
-        lead[i++] = '\0';
-        strcpy(buf, lead);
+    // If we didn't find an exponent marker, then just return the buffer that it is.
+    if (*s != 'E') {
+        buf = val;
+        return;
     }
 
+    // Get the actual exponent, by walking past the eponent marker, then converting to decimal value.
+    int exponent = atoi(++s);
+
+    // Place a NULL char at the exponent marker location, effectively terminating the string there.
     *E = '\0';
-    strcat(buf, v);
+
+    // Now, we setup a (p)ointer to be one byte BEFORE the NULL we just added to the string, this
+    // effectively places us at the end of the numerical string.
+    char *p = E - 1;
+
+    // We're going to start walking from right to left, looking for the last significant digit; the last digit in the string that isn't zero.
+    while ((p >= v) && (*p == '0')) p--;
+
+    // Once found, we can note the position of it in the string.  Since we know that v is a pointer to the start of the string, and p is the last digit found.
+    const int last_significant_digit = (int)(p - v) + 1;
+
+    // Once we know the last siginificant digit, we can calculate how many trailing zero existed.
+    const int trailing_zeros = (int)slen - last_significant_digit;
+    exponent += trailing_zeros;
+    char sbuf[50] = { 0 };
```

Comment:
It would be helpful to rename `sbuf` to `mantissa`. That also resolves the potential confusion with `slen` (which refers to something else).

Reply:
Changed.